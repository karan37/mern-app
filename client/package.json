{
  "name": "tail",
  "version": "0.1.0",
  "private": true,
  "dependencies": {
    "@apollo/client": "^3.11.8",
    "@testing-library/jest-dom": "^5.17.0",
    "@testing-library/react": "^16.0.1",
    "@testing-library/user-event": "^13.5.0",
    "@types/jest": "^27.5.2",
    "@types/node": "^16.18.108",
    "@types/react": "^18.3.5",
    "@types/react-dom": "^18.3.0",
    "antd": "^5.20.6",
    "axios": "^1.7.7",
    "jwt-decode": "^4.0.0",
    "mern-app": "file:..",
    "react": "^18.3.1",
    "react-dom": "^18.3.1",
    "react-router-dom": "^6.26.2",
    "react-scripts": "5.0.1",
    "typescript": "^4.9.5",
    "web-vitals": "^2.1.4"
  },
  "scripts": {
    "start": "react-scripts start",
    "build": "react-scripts build",
    "test": "react-scripts test",
    "eject": "react-scripts eject",
    "format": "prettier --write 'src/**/*.{ts,tsx,js,jsx,json,css,md}'",
    "format:check": "prettier --check 'src/**/*.{ts,tsx,js,jsx,json,css,md}'",
    "lint": "eslint src/**/*.{ts,tsx}",
    "lint:fix": "eslint src/**/*.{ts,tsx} --fix"
  },
  "eslintConfig": {
    "extends": [
      "react-app",
      "react-app/jest"
    ]
  },
  "browserslist": {
    "production": [
      ">0.2%",
      "not dead",
      "not op_mini all"
    ],
    "development": [
      "last 1 chrome version",
      "last 1 firefox version",
      "last 1 safari version"
    ]
  },
  "devDependencies": {
    "@babel/plugin-proposal-private-property-in-object": "^7.21.11",
    "autoprefixer": "^10.4.20",
    "axios-mock-adapter": "^2.0.0",
    "postcss": "^8.4.45",
    "tailwindcss": "^3.4.11",
    "ts-jest": "^29.2.5"
  },
  "jest": {
    "transform": {
      "^.+\\.(ts|tsx)$": "ts-jest"
    },
    "transformIgnorePatterns": [
      "/node_modules/(?!(axios)/)"
    ]
  }
}
