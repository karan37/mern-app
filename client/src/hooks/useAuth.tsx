import { useEffect, useState } from 'react';
import { jwtDecode } from 'jwt-decode';
import { useNavigate } from 'react-router-dom';

interface User {
  id: string;
  name: string;
  role: 'Free User' | 'Paid User';
  iat: number;
  exp: number;
}

const useAuth = () => {
  const [user, setUser] = useState<User | null>(null);
  const navigate = useNavigate();

  useEffect(() => {
    const localToken = localStorage.getItem('token');
    if (localToken) {
      const decodedToken = jwtDecode(localToken) as User;
      const expiry = (decodedToken.exp || 0) * 1000;
      if (Date.now() > expiry) {
        // If token expired
        localStorage.removeItem('token');
        setUser(null);
        navigate('/login');
      } else {
        setUser(decodedToken);
      }
    } else {
      setUser(null);
    }
  }, []);

  const authenticate = (token: string) => {
    localStorage.setItem('token', token);
    setUser(jwtDecode(token));
  };

  const logout = () => {
    localStorage.removeItem('token');
    setUser(null);
    navigate('/login');
  };

  return {
    token: localStorage.getItem('token'),
    isAuthenticated: !!localStorage.getItem('token'),
    authenticate,
    logout,
    user,
    isPaidUser: user?.role === 'Paid User',
  };
};

export default useAuth;
