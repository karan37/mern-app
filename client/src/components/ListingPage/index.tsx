import React, { useState, useEffect } from 'react';
import { Table, Modal, message, Typography } from 'antd';
import { useQuery, gql } from '@apollo/client';
import useAuth from '../../hooks/useAuth';
import Logout from '../LogoutButton';

// GraphQL query to fetch job listings
export const GET_JOB_LISTINGS = gql`
  query GetJobListings {
    jobListings {
      id
      jobTitle
      companyName
      jobType
      jobLevel
      jobGeo
      jobIndustry
      salaryCurrency
    }
  }
`;

const { Title } = Typography;

const ListingPage: React.FC = () => {
  const [isModalOpen, setIsModalOpen] = useState(false);
  const [selectedJob, setSelectedJob] = useState<any>(null);
  const [rows, setRows] = useState(10);
  const { isPaidUser, user } = useAuth();
  const { data, loading, error } = useQuery(GET_JOB_LISTINGS);

  useEffect(() => {
    if (error) {
      message.error('Failed to fetch job listings');
    }
  }, [error]);

  const handleRowClick = (job: any) => {
    setSelectedJob(job);
    setIsModalOpen(true);
  };

  const columns = [
    { title: 'Job Title', dataIndex: 'jobTitle', key: 'title' },
    { title: 'Company Name', dataIndex: 'companyName', key: 'companyName' },
    { title: 'Job Type', dataIndex: 'jobType', key: 'jobType' },
    { title: 'Job Level', dataIndex: 'jobLevel', key: 'jobLevel' },
  ];

  return (
    <>
      <div>
        <Title level={2} className="float-left">
          Welcome to the job board{' '}
          <span className="text-main-blue">{user?.name}</span>!
        </Title>
        <Logout className="float-right mt-1" />
      </div>
      <Table
        dataSource={data?.jobListings || []}
        loading={loading}
        columns={columns}
        rowKey="id"
        className="[&_.ant-table-thead_th]:!bg-main-blue [&_.ant-table-thead_th]:!text-white"
        onRow={(record) => ({
          onClick: () => handleRowClick(record),
          className: 'cursor-pointer',
        })}
        scroll={{ y: 'calc(100vh - 200px)' }}
        pagination={{
          pageSize: rows,
          pageSizeOptions: isPaidUser ? ['10', '20', '50'] : ['10'], // Options for Paid User, default to 10 for Free Users
          showSizeChanger: isPaidUser, // Allow only Paid Users to change page size
          onChange: (page, pageSize) => setRows(pageSize || 10),
        }}
        rowClassName={(record, index) =>
          index % 2 === 0 ? 'bg-gray-300' : 'bg-gray-200'
        }
      />
      {selectedJob && (
        <Modal
          title={
            isPaidUser ? 'Job Details' : 'Feature Not Available For Free Users'
          }
          open={isModalOpen}
          onCancel={() => setIsModalOpen(false)}
          footer={null}
        >
          <p>
            <strong>Job Geo:</strong>{' '}
            {isPaidUser ? selectedJob.jobGeo : '?????'}
          </p>
          <p>
            <strong>Salary Currency:</strong>{' '}
            {isPaidUser ? selectedJob.salaryCurrency : '?????'}
          </p>
          <p>
            <strong>Job Industry:</strong>{' '}
            {isPaidUser ? selectedJob.jobIndustry.join(', ') : '?????'}
          </p>
        </Modal>
      )}
    </>
  );
};

export default ListingPage;
