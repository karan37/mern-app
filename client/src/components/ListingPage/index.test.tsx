import { render, screen, fireEvent, act, within } from '@testing-library/react';
import ListingPage from '.';
import { message } from 'antd';
import useAuth from '../../hooks/useAuth';
import { useNavigate } from 'react-router-dom';
import { useQuery } from '@apollo/client';

jest.mock('../../hooks/useAuth');

jest.mock('antd', () => ({
  ...jest.requireActual('antd'),
  message: { error: jest.fn(), success: jest.fn() },
}));

jest.mock('@apollo/client', () => ({
  gql: jest.fn(),
  useQuery: jest.fn(),
}));

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: jest.fn(),
}));

const mockNavigate = useNavigate as jest.MockedFunction<typeof useNavigate>;
const mockUseQuery = useQuery as jest.Mock;
const mockUseAuth = useAuth as jest.Mock;

describe('ListingPage Component', () => {
  const jobListingsMock = [
    {
      id: '1',
      jobTitle: 'Software Engineer',
      companyName: 'Tech Corp',
      jobType: 'Full-time',
      jobLevel: 'Senior',
      jobGeo: 'New York',
      salaryCurrency: 'USD',
      jobIndustry: ['Engineering'],
    },
  ];

  it('renders job listings and allows row click to show modal for Paid User', async () => {
    mockUseAuth.mockReturnValue({
      isPaidUser: true,
      token: 'mockToken',
    });
    mockUseQuery.mockReturnValue({
      data: {
        jobListings: jobListingsMock,
      },
      loading: false,
      error: null,
    });

    render(<ListingPage />);

    // Check if the table is populated
    expect(screen.getByText('Software Engineer')).toBeInTheDocument();
    expect(screen.getByText('Tech Corp')).toBeInTheDocument();

    // Simulate a row click to show modal
    await act(() => {
      fireEvent.click(screen.getByText('Software Engineer'));
    });

    // Check if the modal appears with full job details for Paid User
    expect(screen.getByText(/New York/i)).toBeInTheDocument;
    expect(screen.getByText(/USD/i)).toBeInTheDocument;
    expect(screen.getByText(/Engineering/i)).toBeInTheDocument;
  });

  it('renders error message on query error', async () => {
    mockUseAuth.mockReturnValue({
      isPaidUser: false,
      token: 'mockToken',
    });
    mockUseQuery.mockReturnValue({
      data: null,
      loading: false,
      error: new Error('GraphQL error'),
    });
    render(<ListingPage />);

    // Check if error message is displayed
    expect(message.error).toHaveBeenCalledWith('Failed to fetch job listings');
  });

  it('restricts Paid User features for Free User', async () => {
    mockUseAuth.mockReturnValue({
      isPaidUser: false,
      token: 'mockToken',
    });
    mockUseQuery.mockReturnValue({
      data: {
        jobListings: jobListingsMock,
      },
      loading: false,
      error: null,
    });

    render(<ListingPage />);

    // Check if the table is populated
    expect(screen.getByText('Software Engineer')).toBeInTheDocument();

    // Simulate a row click to show modal
    await act(async () => {
      fireEvent.click(screen.getByText('Software Engineer'));
    });

    // Check if the modal appears with restricted details for Free User
    const jobGEo = screen.getByText(/Job Geo/i).closest('p');
    expect(within(jobGEo!).getByText('?????')).toBeInTheDocument;
    const salaryCurrency = screen.getByText(/Salary Currency/i).closest('p');
    expect(within(salaryCurrency!).getByText('?????')).toBeInTheDocument;
    const jobIndustry = screen.getByText(/Job Industry/i).closest('p');
    expect(within(jobIndustry!).getByText('?????')).toBeInTheDocument;
    expect(screen.getByText(/Feature Not Available For Free Users/i))
      .toBeInTheDocument;
    expect(salaryCurrency).toMatchSnapshot();
    expect(jobGEo).toMatchSnapshot();
    expect(jobIndustry).toMatchSnapshot();
  });

  it('logs out and redirects to login page on Logout button click', async () => {
    const mockLogout = jest.fn();
    mockUseAuth.mockReturnValue({
      isPaidUser: 'Free User',
      logout: mockLogout,
    });
    mockUseQuery.mockReturnValue({
      data: {
        jobListings: jobListingsMock,
      },
      loading: false,
      error: null,
    });

    const navigate = jest.fn();
    mockNavigate.mockReturnValue(navigate);

    render(<ListingPage />);

    const logoutButton = screen.getByText('Logout');

    // Click Logout
    fireEvent.click(logoutButton);

    expect(mockLogout).toHaveBeenCalled();
  });
});
