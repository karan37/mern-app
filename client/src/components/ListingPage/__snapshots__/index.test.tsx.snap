// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListingPage Component restricts Paid User features for Free User 1`] = `
<p>
  <strong>
    Salary Currency:
  </strong>
   
  ?????
</p>
`;

exports[`ListingPage Component restricts Paid User features for Free User 2`] = `
<p>
  <strong>
    Job Geo:
  </strong>
   
  ?????
</p>
`;

exports[`ListingPage Component restricts Paid User features for Free User 3`] = `
<p>
  <strong>
    Job Industry:
  </strong>
   
  ?????
</p>
`;
